include/aws/common/cbor.h

```diff
@@ -0,0 +1,450 @@
+
+/**
+ * @brief Encode a AWS_CBOR_TYPE_ARRAY_START value to "smallest possible" in encoder's buffer.
+ *  Referring to RFC8949 section 4.2.1
+ *  The "number_entries" is the cbor data items should be followed as the content of the array.
+ * Notes: it's user's responsibility to keep the integrity of the array to be encoded.
+ *
+ * @param encoder
+ * @param number_entries The number of data item in array.
+ */
+AWS_COMMON_API
+void aws_cbor_encoder_write_array_start(struct aws_cbor_encoder *encoder, size_t number_entries);
+
+/**
+ * @brief Encode a AWS_CBOR_TYPE_MAP_START value to "smallest possible" in encoder's buffer.
+ *  Referring to RFC8949 section 4.2.1
+ *  The "number_entries" is the number of pair of cbor data items as key and value should be followed as the content of
+ * the map.
+ *
+ * Notes: it's user's responsibility to keep the integrity of the map to be encoded.
+ *
+ * @param encoder
+ * @param number_entries The number of data item in map.
+ */
+AWS_COMMON_API
+void aws_cbor_encoder_write_map_start(struct aws_cbor_encoder *encoder, size_t number_entries);
+
+/**
+ * @brief Encode a AWS_CBOR_TYPE_TAG value to "smallest possible" in encoder's buffer.
+ *  Referring to RFC8949 section 4.2.1
+ * The following cbor data item will be the content of the tagged value.
+ * Notes: it's user's responsibility to keep the integrity of the tagged value to follow the RFC8949 section 3.4
+ *
+ * @param encoder
+ * @param tag_number The tag value to encode.
+ */
+AWS_COMMON_API
+void aws_cbor_encoder_write_tag(struct aws_cbor_encoder *encoder, uint64_t tag_number);
+
+/**
+ * @brief Encode a simple value AWS_CBOR_TYPE_NULL
+ *
+ * @param encoder
+ */
+AWS_COMMON_API
+void aws_cbor_encoder_write_null(struct aws_cbor_encoder *encoder);
+
+/**
+ * @brief Encode a simple value AWS_CBOR_TYPE_UNDEFINE
+ *
+ * @param encoder
+ */
+AWS_COMMON_API
+void aws_cbor_encoder_write_undefine(struct aws_cbor_encoder *encoder);
+
+/**
+ * @brief Encode a simple value AWS_CBOR_TYPE_BOOL
+ *
+ * @param encoder
+ */
+AWS_COMMON_API
+void aws_cbor_encoder_write_bool(struct aws_cbor_encoder *encoder, bool value);
+
+/**
+ * @brief Encode a simple value AWS_CBOR_TYPE_BREAK
+ *
+ * Notes: no error checking, it's user's responsibility to track the break
+ * to close the corresponding inf_start
+ */
+AWS_COMMON_API
+void aws_cbor_encoder_write_break(struct aws_cbor_encoder *encoder);
+
+/**
+ * @brief Encode inf start, supported type is:
+ *  AWS_CBOR_TYPE_INF_BYTESTRING_START
+ *  AWS_CBOR_TYPE_INF_STRING_START
+ *  AWS_CBOR_TYPE_INF_ARRAY_START
+ *  AWS_CBOR_TYPE_INF_MAP_START
+ * All other type will result in error.
+ *
+ * Notes: It's user's responsibility to track and add the break to
+ * close the corresponding inf_start
+ *
+ * @param encoder
+ * @param type
+ */
+AWS_COMMON_API
+int aws_cbor_encoder_write_inf_start(struct aws_cbor_encoder *encoder, enum aws_cbor_element_type type);
```

Comment:
This is the only write() function that isn't `void`

I'd suggest keeping it simple/consistent by just having 4 separate functions for bytes/text/array/map

Or just naively encode the bad type and continue on our merry way, since there are plenty of other ways for users to encode data illegally